src/Interpreters/ProcessList.cpp

```diff
@@ -373,16 +377,43 @@ QueryStatus::~QueryStatus()
     }
 }
 
+void QueryStatus::ExecutorHolder::cancel()
+{
+    std::lock_guard lock(mutex);
+    if (executor)
+        executor->cancel();
+}
+
+void QueryStatus::ExecutorHolder::remove()
+{
+    std::lock_guard lock(mutex);
+    executor = nullptr;
+}
+
 CancellationCode QueryStatus::cancelQuery(bool)
 {
     if (is_killed.load())
         return CancellationCode::CancelSent;
 
     is_killed.store(true);
 
-    std::lock_guard lock(executors_mutex);
-    for (auto * e : executors)
+    std::unique_lock lock(executors_mutex);
+    for (const auto & e : executors)
+    {
+        /// We should call cancel() with unlocked executors_mutex, because
+        /// cancel() can try to lock some internal mutex that is already locked by query executing
+        /// thread, and query executing thread can call removePipelineExecutor and lock executors_mutex,
+        /// which will lead to deadlock.
+        /// Note that the size and the content of executors cannot be changed while
+        /// executors_mutex is unlocked, because:
+        /// 1) We don't allow adding new executors while cancelling query in addPipelineExecutor
+        /// 2) We don't actually remove executor holder from executors in removePipelineExecutor,
+        /// just mark that executor is invalid.
+        /// So, it's safe to continue iteration over executors after subsequent mutex locking.
+        lock.unlock();
```

Comment:
Maybe better not to perform unlock/lock on each iteration, but to collect `executers` snapshot and then for each element call `cancel`?

Reply:
Done